### PR TITLE
SMSService

### PR DESCRIPTION
**Pull request by blakegreen32**
Implemented SMS services and DAOs using Twilio to send SMS notifications.

Ticket: https://nanopay.atlassian.net/browse/NP-1206
nanopay PR: https://github.com/nanoPayinc/NANOPAY/pull/9961

### CODE REVIEW

File: src/foam/nanos/notification/sms/NullSMSService.js [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+  implements: [
+    'foam.nanos.notification.sms.SMSService'
+  ],
+
+  methods: [
+    {
+      name: 'sendSms',
+      javaCode: 'return smsMessage;',
+      code: function() {
+        return;
```
Comment (kgrgreer) [inline src/foam/nanos/notification/sms/NullSMSService.js]: This line is redundant. Just add a line like this to indicate you intend to do nothing (NOP = No-Operation):
// NOP

File: src/foam/nanos/notification/sms/ProxySMSService.js [hunk 2]
```diff
@@ -0,0 +1,24 @@
+/**
+ * @license
+ * Copyright 2019 The FOAM Authors. All Rights Reserved.
```
Comment (kgrgreer) [inline src/foam/nanos/notification/sms/ProxySMSService.js]: 2020

File: src/foam/nanos/notification/sms/TwilioSMSService.js [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+  name: 'TwilioSMSService',
+
+  implements: [
+    'foam.nanos.notification.sms.SMSService'
+  ],
+
+  documentation: 'Sends SMS message through Twilio',
+
+  javaImports: [
+    'com.twilio.Twilio',
```
Comment (kgrgreer) [inline src/foam/nanos/notification/sms/TwilioSMSService.js]: sort imports

File: src/foam/nanos/notification/services [hunk 4]
```diff
@@ -113,3 +113,51 @@ p({
   "serviceClass":"foam.nanos.notification.email.POP3EmailService",
   "client":"{ \"class\":\"foam.nanos.notification.email.ClientPOP3EmailService\", \"delegate\": { \"class\":\"foam.box.HTTPBox\", \"url\":\"service/pop3\" } }"
 })
+
+p({
+  "class":"foam.nanos.boot.NSpec",
+  "name":"localSmsMessageDAO",
+  "serviceScript":"""
+    return new foam.dao.EasyDAO.Builder(x)
```
Comment (jlhughes) [inline src/foam/nanos/notification/services:121]: Please set this up like the emailMessageDAO - with fixed sized support so we don't load all SMSs into memory.

File: src/foam/nanos/notification/services [hunk 5]
```diff
@@ -113,3 +113,51 @@ p({
...
+})
+
+p({
+  "class":"foam.nanos.boot.NSpec",
+  "name":"smsMessageDAO",
+  "serve":true,
+  "serviceScript":"""
+    return new foam.dao.EasyDAO.Builder(x)
+      .setOf(foam.nanos.notification.sms.SMSMessage.getOwnClassInfo())
+      .setInnerDAO((foam.dao.DAO) x.get(\"localSmsMessageDAO\"))
```
Comment (jlhughes) [inline src/foam/nanos/notification/services]: all the \\" are not required when the script is inside the """

File: src/foam/nanos/notification/sms/NullSMSService.js [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+  package: 'foam.nanos.notification.sms',
+  name: 'NullSMSService',
+
+  implements: [
+    'foam.nanos.notification.sms.SMSService'
+  ],
+
+  methods: [
+    {
+      name: 'sendSms',
```
Comment (jlhughes) [inline src/foam/nanos/notification/sms/NullSMSService.js]: was 'send' already used?

File: src/foam/nanos/notification/sms/TwilioSMSService.js [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+    'com.twilio.type.PhoneNumber',
+    'foam.dao.DAO',
+    'foam.nanos.app.TwilioConfig',
+    'foam.nanos.notification.sms.SMSStatus',
+    'foam.util.SafetyUtil'
+  ],
+
+  methods: [
+    {
+      name: 'sendSms',
```
Comment (jlhughes) [inline src/foam/nanos/notification/sms/TwilioSMSService.js]: Can you add OMs similar SMTPservice please then we can alarm on congestion and failure.

Comment (blakegreen32) [reply src/foam/nanos/notification/sms/TwilioSMSService.js:29]: Guessing you meant PMs, ya sure

File: src/foam/nanos/notification/services (hunk 5 above)
Comment (nanoscott) [reply src/foam/nanos/notification/services]: you mean backslashes aren't required, right?

File: src/foam/nanos/notification/sms/NullSMSService.js (hunk 1 above)
Comment (nanoscott) [inline src/foam/nanos/notification/sms/NullSMSService.js]:
```suggestion
        // NOP
```

File: src/foam/nanos/notification/sms/TwilioSMSService.js [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+        SMSStatus status;
+
+        // check if phone number exists
+        if ( ! SafetyUtil.isEmpty(smsMessage.getPhoneNumber()) ) {
+          phoneNumber = smsMessage.getPhoneNumber();
+        } else {
+          status = SMSStatus.UNSENT;
+          status.setErrorMessage("Phone number not found, cannot send SMS.");
+          smsMessage.setStatus(status);
+          return (SMSMessage) smsMessageDAO.put(smsMessage);
```
Comment (nanoscott) [inline src/foam/nanos/notification/sms/TwilioSMSService.js]: if we are strictly modelling against the email system, then you shouldn't be putting the smsMessage to the DAO here, but instead you should be letting the cron job or script or calling code put the returned smsMessage to the DAO .. keep this simple and just let it do the sending and set the status and allow the calling code to do the coordination.

File: src/foam/nanos/notification/sms/TwilioSMSService.js [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+        Twilio.init(twilioConfig.getAccountSid(), twilioConfig.getAuthToken());
+
+        String phoneNumber = "";
+        SMSStatus status;
+
+        // check if phone number exists
+        if ( ! SafetyUtil.isEmpty(smsMessage.getPhoneNumber()) ) {
+          phoneNumber = smsMessage.getPhoneNumber();
+        } else {
+          status = SMSStatus.UNSENT;
```
Comment (nanoscott) [inline src/foam/nanos/notification/sms/TwilioSMSService.js]: this should be FAILED, UNSENT should be saved for those messages that still need to be sent.

File: src/foam/nanos/notification/sms/TwilioSMSService.js [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+          status.setErrorMessage("Phone number not found, cannot send SMS.");
+          smsMessage.setStatus(status);
+          return (SMSMessage) smsMessageDAO.put(smsMessage);
+        }
+
+        if ( ! SafetyUtil.isEmpty(smsMessage.getMessage()) ) {
+          try {
+            Message.creator(new PhoneNumber(phoneNumber), new PhoneNumber(twilioConfig.getPhoneNumber()),
```
Comment (nanoscott) [inline src/foam/nanos/notification/sms/TwilioSMSService.js:50]: is anythign returned from the Twilio code to indicate that the call was successful? Or are we just relying on exceptions to indicate failure?

File: src/foam/nanos/notification/services (hunk 5 above)
Comment (jlhughes) [reply src/foam/nanos/notification/services]: yes.

File: src/foam/nanos/notification/sms/NullSMSService.js (hunk 6 above)
Comment (nanoscott) [inline src/foam/nanos/notification/sms/NullSMSService.js]:
```suggestion
      name: 'send',
```

File: src/foam/nanos/notification/sms/SMSService.js [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+ foam.INTERFACE({
+   package: 'foam.nanos.notification.sms',
+   name: 'SMSService',
+
+   methods: [
+     {
+       name: 'sendSms',
```
Comment (nanoscott) [inline src/foam/nanos/notification/sms/SMSService.js]:
```suggestion
       name: 'send',
```

File: src/foam/nanos/notification/sms/TwilioSMSService.js [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+    'com.twilio.type.PhoneNumber',
+    'foam.nanos.app.TwilioConfig',
+    'foam.nanos.notification.sms.SMSStatus',
+    'foam.nanos.om.OMLogger',
+    'foam.util.SafetyUtil'
+  ],
+
+  methods: [
+    {
+      name: 'sendSms',
```
Comment (nanoscott) [inline src/foam/nanos/notification/sms/TwilioSMSService.js]:
```suggestion
      name: 'send',
```

File: src/foam/nanos/notification/sms/TwilioSMSService.js [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+      javaCode: `
+        OMLogger omLogger = (OMLogger) x.get("OMLogger");
+        TwilioConfig twilioConfig = (TwilioConfig) x.get("twilioConfig");
+        Twilio.init(twilioConfig.getAccountSid(), twilioConfig.getAuthToken());
```
Comment (nanoscott) [inline src/foam/nanos/notification/sms/TwilioSMSService.js]: validate that the SID and token are set .. if they are not, log an error message and return right away, leaving the message unsent.

File: src/foam/nanos/notification/services [hunk 14]
```diff
@@ -113,3 +113,54 @@ p({
...
+p({
+  "class":"foam.nanos.boot.NSpec",
+  "name":"localSmsMessageDAO",
+  "serviceScript":"""
+    return new foam.dao.EasyDAO.Builder(x)
+      .setOf(foam.nanos.notification.sms.SMSMessage.getOwnClassInfo())
+      .setAuthorize(false)
+      .setSeqNo(true)
+      .setJournalType(foam.dao.JournalType.SINGLE_JOURNAL)
+      .setJournalName(\"smsMessages\")
```
Comment (jlhughes) [inline src/foam/nanos/notification/services]:
```suggestion
      .setJournalName("smsMessages")
```

File: src/foam/nanos/notification/sms/TwilioSMSService.js [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+          return smsMessage;
+        }
+
+        // check if message exists
+        if ( ! SafetyUtil.isEmpty(smsMessage.getMessage()) ) {
+          try {
+            Message.creator(new PhoneNumber(phoneNumber), new PhoneNumber(twilioConfig.getPhoneNumber()),
+            smsMessage.getMessage()).create();
+            smsMessage.setStatus(SMSStatus.SENT);
+            omLogger.log(this.getClass().getSimpleName(), "message", "sent");
```
Comment (jlhughes) [inline src/foam/nanos/notification/sms/TwilioSMSService.js:65]: There needs to be another before with 'sending'. The alarming system works by matching OMs. so if get a log of sending, and no 'sent', then we know we have congestion.

Comment (jlhughes) [reply src/foam/nanos/notification/sms/TwilioSMSService.js:65]: See SMTPEmailService transport_.  'connecting'

**Review by jlhughes — APPROVED** [review]

**Review by nanoscott — APPROVED** [review]

**Pull request merged**